File: 784-LetterCasePermutation/784-LetterCasePermutation.java
/*

784. Letter Case Permutation
Easy

Given a string S, we can transform every letter individually to be lowercase or uppercase to create another string.  Return a list of all possible strings we could create.

Examples:
Input: S = "a1b2"
Output: ["a1b2", "a1B2", "A1b2", "A1B2"]

Input: S = "3z4"
Output: ["3z4", "3Z4"]

Input: S = "12345"
Output: ["12345"]

*/

class Solution {
    public List<String> letterCasePermutation(String S) {
        List<String> res = new ArrayList<>();
        if(S.length() == 0) return res;
        dfs(S, res, 0, "");
        return res;
    }
    
    private void dfs(String S, List<String> res, int curr, String s){
        if (s.length() == S.length()){
            res.add(s);
            return;
        }
        char ch = S.charAt(curr);
        if(ch >= '0' && ch <= '9'){
            dfs(S, res, curr + 1, s + String.valueOf(ch));
            return;
        }
         dfs(S, res, curr + 1, s + String.valueOf(Character.toUpperCase(ch)));
         dfs(S, res, curr + 1, s + String.valueOf(Character.toLowerCase(ch)));
    }
}

File: 784-LetterCasePermutation/784-LetterCasePermutation.py
"""

784. Letter Case Permutation

Given a string s, you can transform every letter individually to be lowercase or uppercase to create another string.

Return a list of all possible strings we could create. Return the output in any order.

Example 1:

Input: s = "a1b2"
Output: ["a1b2","a1B2","A1b2","A1B2"]

Example 2:

Input: s = "3z4"
Output: ["3z4","3Z4"]

Constraints:

    1 <= s.length <= 12
    s consists of lowercase English letters, uppercase English letters, and digits.

"""

class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        def dfs(idx, curr):
            if len(curr) == len(s):
                res.append(curr)
                return
            if s[idx] >= '0' and s[idx] <= '9':
                dfs(idx + 1, curr + s[idx])
                return
            dfs(idx + 1, curr + s[idx].lower())
            dfs(idx + 1, curr + s[idx].upper())
        res = []
        dfs(0, "")
        return res
